rfcs/0000-state-separation.md

```diff
@@ -0,0 +1,706 @@
+# Table of Contents
+
+- [Tracking Issues](#tracking-issue)
+  - [Amethyst Community Forum Discussion](#forum-discussion)
+- [Motivation]
+- [Guide Level Explanation](#guide-level-explanation)
+- [Reference Level Explanation](#reference-level-explanation)
+- [Drawbacks]
+- [Rationale and Alternatives](#rationale-and-alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+
+# Basic Info
+[basic]: #basic-info
+
+- Feature Name: state_separation
+- Start Date: 2018-11-30
+- RFC PR: (leave this empty until a PR is opened)
+- [Tracking Issue](#tracking-issue): (leave this empty)
+- [Forum Thread](#forum-discussion)
+
+# Summary
+[summary]: #summary
+
+Create a clean separation between states, and the callbacks (code) associated with them.
+
+## Amethyst Community Forum Discussion
+[forum-discussion]: #forum-discussion
+
+https://community.amethyst-engine.org/t/request-for-comments-simplified-state-machine/143/3
+
+# Motivation
+[motivation]: #motivation
+
+* Improve ergonomics and re-usability of state code.
+* Track the current state in a manner which is available through the `World`.
+
+# Guide-Level Explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Here's an overview of what I'm proposing, followed by dedicated sections where I try to walk
+through how each change works and what motivated them:
+
+* States and their callbacks are separated.
+* Introducing a `State` derive to construct highly efficient storage for state enums.
+* Getting the current state from a system.
+* Introducing global callbacks.
+* Custom GameData in states is deprecated.
+* Modifying states is done through a `States` resource.
+* Refactor amethyst_test.
+
+Finally there will be an example application as it would be written incorporating these changes.
+
+## States and their callbacks are separated
+
+State implements the `trait State<T, E: Send + Sync + 'static>` trait.
+
+T is very commonly some variation of `GameData<'a, 'b>`. This can be customized (see below), but the only real requirement is that it somehow encapsulates a `Dispatcher`.
+Every state is then responsible for driving the dispatcher through their `update` function.
+
+This is primarily done by implementing `SimpleState`, and then relying on the blanket implementation of this to perform the dispatch.
+
+This change instead introduces the following:
+
+* A simplified `StateCallback<S, E>` trait, where `S: State<E>`, and `E` is the event.
```

Comment:
I'm not a huge fan of the use of `Callback`, maybe `StateHandler` might be better?

Reply:
`StateHandler` is a good alternative. Naming is hard, so feel free to suggest better ones for anything.

Reply:
Callback has been changed to Handler, kudos!